apps/wallet-mobile/src/features/Swap/common/SwapFormProvider.tsx

```diff
@@ -1,86 +1,415 @@
-      throw new Error(`touchedReducer invalid action`)
+        break
+
+      case SwapFormActionType.ResetSwapForm:
+        return defaultState
+
+      case SwapFormActionType.CanSwapChanged:
+        draft.canSwap = action.canSwap
+
+        break
+
+      case SwapFormActionType.SellInputValueChanged:
+        draft.sellAmount.displayValue = action.value
+
+        break
+
+      case SwapFormActionType.BuyInputValueChanged:
+        draft.buyAmount.displayValue = action.value
+
+        break
+
+      case SwapFormActionType.LimitInputValueChanged:
+        draft.limitAmount.displayValue = action.value
+
+        break
+
+      case SwapFormActionType.SellAmountErrorChanged:
+        draft.sellAmount.error = action.error
+
+        break
+
+      case SwapFormActionType.BuyAmountErrorChanged:
+        draft.buyAmount.error = action.error
+
+        break
+
+      default:
+        throw new Error(`touchedReducer invalid action`)
+    }
+  })
+}
+
+const PRECISION = 14
+
+const defaultState: SwapFormState = Object.freeze({
+  sellAmount: {
+    isTouched: true,
+    disabled: false,
+    error: undefined,
+    displayValue: '',
+  },
+  buyAmount: {
+    isTouched: false,
+    disabled: false,
+    error: undefined,
+    displayValue: '',
+  },
+  selectedPool: {
+    isTouched: false,
+  },
+  limitAmount: {
+    displayValue: '',
+  },
+  canSwap: false,
+})
+
+/* istanbul ignore next */
+function missingInit() {
+  console.error('[SwapFormContext] missing initialization')
+}
+
+const initialSwapFormContext: SwapFormContext = {
+  ...defaultState,
+  sellTouched: missingInit,
+  buyTouched: missingInit,
+  poolTouched: missingInit,
+  poolDefaulted: missingInit,
+  switchTouched: missingInit,
+  switchTokens: missingInit,
+  resetSwapForm: missingInit,
+  buyInputValueChanged: missingInit,
+  sellInputValueChanged: missingInit,
+  limitInputValueChanged: missingInit,
+  buyAmountErrorChanged: missingInit,
+  sellAmountErrorChanged: missingInit,
+  canSwapChanged: missingInit,
+  sellInputRef: undefined,
+  buyInputRef: undefined,
+  limitInputRef: undefined,
+  onChangeSellQuantity: missingInit,
+  onChangeBuyQuantity: missingInit,
+  onChangeLimitPrice: missingInit,
+}
+
+type SwapFormAction =
+  | {type: SwapFormActionType.SellTouched}
+  | {type: SwapFormActionType.BuyTouched}
+  | {type: SwapFormActionType.SwitchTouched}
+  | {type: SwapFormActionType.PoolTouched}
+  | {type: SwapFormActionType.PoolDefaulted}
+  | {type: SwapFormActionType.ResetSwapForm}
+  | {type: SwapFormActionType.CanSwapChanged; canSwap: boolean}
+  | {type: SwapFormActionType.SellInputValueChanged; value: string}
+  | {type: SwapFormActionType.BuyInputValueChanged; value: string}
+  | {type: SwapFormActionType.SellAmountErrorChanged; error: string | undefined}
+  | {type: SwapFormActionType.LimitInputValueChanged; value: string}
+  | {type: SwapFormActionType.BuyAmountErrorChanged; error: string | undefined}
+
+type SwapFormState = {
+  sellAmount: {
```

Comment:
hmmm, @javi I think these are sellQuantity and buyQuantity, we save Amount for token+quantity, and here it is representing the quantity input only I think we should refactor to keep consistency